core/src/main/java/com/linecorp/armeria/unsafe/PooledObjects.java

```diff
@@ -148,6 +149,36 @@ public static <T> T copyAndClose(T obj) {
                 }
             }
         }
+        if (obj instanceof WebSocketFrame) {
+            final WebSocketFrame frame = (WebSocketFrame) obj;
+            if (frame.isPooled()) {
+                try {
+                    switch (frame.type()) {
+                        case CONTINUATION:
+                            //noinspection unchecked
+                            return (T) WebSocketFrame.ofContinuation(frame.array(), frame.isFinalFragment());
+                        case TEXT:
+                            //noinspection unchecked
+                            return (T) WebSocketFrame.ofText(frame.array(), frame.isFinalFragment());
+                        case BINARY:
+                            //noinspection unchecked
+                            return (T) WebSocketFrame.ofBinary(frame.array(), frame.isFinalFragment());
+                        case CLOSE:
+                            //noinspection unchecked
+                            return (T) WebSocketFrame.ofClose(frame.array());
+                        case PING:
+                            //noinspection unchecked
+                            return (T) WebSocketFrame.ofPing(frame.array());
+                        case PONG:
+                            //noinspection unchecked
+                            return (T) WebSocketFrame.ofPong(frame.array());
+                    }
+                } finally {
+                    frame.close();
+                }
```

Comment:
Could extract this into a method to use `@SuppressWarnings` instead of repeating line comments.